### ISSUE TITLE
Bookmarks garbled

### ISSUE THREAD

**Issue by ipk01** [xbrowsersync/app#89]
### Steps to reproduce
1. Sync on various machines and architectures
2. 
3.

### Expected behaviour
Bookmarks should be the same on all machines

### Actual behaviour
xBrowsersync generated a duplicate folder in my bookmarks bar where only 2 new bookmarks had been added to a subfolder of another folder.
i.e.: I have added bookmarks x and y to folder a/b/c
xBrowsersync added a copy of the complete folder tree, i.e. a'/b'/c' and put the new bookmarks there.
When trying to verify on a third machine, it dropped70% of my bookmarks alltogether after synching.
-> Bookmarks completely corrupted on all subsequent syncs.

### Client app information
**App Platform:** (Chrome, Firefox, etc) 
Chrome (no Firefox involved), latest pre_release of the App (v6).

**App Platform Version:** 
Last intact version on Ubuntu on my home PC, first corruption noticed on Win10 at work, complete destruction then on Ubuntu.

#### Other info
```
Enter any other information you think would be helpful here.
```

**Comment by nero120** [comment]
Did this occur with the beta on all of your machines? If so can you provide logs?

**Comment by ipk01** [comment]
I came into work this morning to find my bookmarks garbled after attempting to sync and the sync process hanging (spinning wheel never stopped). This MAY be related to me using the "test" server (which is default in v6, and i forgot to change that when I created a new sync at work yesterday). In any case: Here is the log. I can provide the log from my home machine tonight.
20190506142254434	[trace]	Starting up
20190506142254442	[trace]	{"displayPermissions":true,"appVersion":"1.5.0beta6","platform":{"name":"chrome","version":"74.0.3729","mobile":false,"os":"Windows 10"}}
20190506142643165	[trace]	Sync izsr45 (push) queued (1 waiting to sync)
20190506142643581	[trace]	Sync enabled
20190506142643584	[trace]	Sync izsr45 completed (0 waiting to sync)
20190506142643590	[trace]	New sync id created: 4717c07a4f1741889a1d32b506aba868
20190507080053860	[trace]	Starting up
20190507080054066	[trace]	{"displayIntro":false,"displayPermissions":false,"lastUpdated":"2019-05-06T12:26:54.014Z","networkDisconnected":false,"syncEnabled":true,"syncId":"4717c07a4f1741889a1d32b506aba868","syncVersion":"1.5.0beta6","appVersion":"1.5.0beta6","platform":{"name":"chrome","version":"74.0.3729","mobile":false,"os":"Windows 10"}}
20190507080054769	[trace]	Updates found, retrieving latest sync data
20190507080054771	[trace]	Updates found, retrieving latest sync data
20190507080054773	[trace]	Sync gpbuhz (pull) queued (1 waiting to sync)
20190507080055124	[trace]	Local population completed in 0.23s
20190507080055135	[trace]	Sync 11h4ml4 completed (1 waiting to sync)
20190507080055297	[trace]	onRemoved event detected
20190507080055302	[trace]	Sync 1cgs51z (push) queued (15 waiting to sync)
20190507080055308	[trace]	onCreated event detected
20190507080055316	[trace]	Sync 1im6c2e (push) queued (16 waiting to sync)
20190507080055328	[trace]	Sync 16evbjv (push) queued (17 waiting to sync)
20190507080055335	[trace]	Sync 1f8m9ra (push) queued (18 waiting to sync)
20190507080055340	[trace]	Sync 1amkre2 (push) queued (19 waiting to sync)
20190507080055345	[trace]	Sync bi82y1 (push) queued (20 waiting to sync)
20190507080055420	[trace]	Sync 9pjb3s (push) queued (21 waiting to sync)
20190507080056057	[trace]	onCreated event detected
20190507080056243	[trace]	Sync 1y2w33l (push) queued (45 waiting to sync)
20190507080056355	[trace]	Sync 16es7u2 (push) queued (200 waiting to sync)
20190507080056398	[trace]	Sync gpbuhz completed (278 waiting to sync)
20190507080056433	[trace]	Sync 10wze10 (push) queued (281 waiting to sync)
20190507080056458	[trace]	Sync 6z3swz (push) queued (282 waiting to sync)
20190507080057449	[trace]	onRemoved event detected
20190507080057730	[trace]	Sync xvfk13 (push) queued (448 waiting to sync)
20190507080058058	[trace]	Sync mlcml3 (push) queued (984 waiting to sync)
20190507080058336	[trace]	Sync 1ie3xzn (push) queued (1220 waiting to sync)
20190507080106138	[trace]	Waiting for sync: 1ie3xzn
20190507080325333	[trace]	Waiting for sync: 1ie3xzn
20190507080402566	[trace]	Sync disabled

**Comment by nero120** [comment]
Thanks for this @ipk01, looks like the event handlers were not disabled when syncing updates. I'll dig into it.

**Comment by ipk01** [comment]
Hmmm ... I cannot recall ... but it may be I triggered the last sync by clicking on the icon, in case that helps generate ideas.

**Comment by ipk01** [comment]
Here is the log from the home machine, by the way ... Last time I left the machine, the bookmarks were fine, but they also got garbled now today.
20190506210650581	[trace]	Starting up
20190506210650593	[trace]	{"displayPermissions":true,"appVersion":"1.5.0beta6","platform":{"name":"chrome","version":"74.0.3729","mobile":false,"os":"Linux x86_64"}}
20190506210655977	[trace]	Optional permissions granted
20190506210707959	[trace]	Sync 1f2pr30 (pull) queued (1 waiting to sync)
20190506210708859	[trace]	Local population completed in 0.72s
20190506210708871	[trace]	Sync enabled
20190506210709046	[trace]	Synced to existing id: 4717c07a4f1741889a1d32b506aba868
20190506210733255	[trace]	onRemoved event detected
20190506210733283	[trace]	Sync knd30r (push) queued (1 waiting to sync)
20190506210734668	[trace]	Sync knd30r completed (0 waiting to sync)
20190506231455281	[trace]	onCreated event detected
20190506231455342	[trace]	Sync 1z0qvpo (push) queued (1 waiting to sync)
20190506231455974	[trace]	Sync 1z0qvpo completed (0 waiting to sync)
20190506231504185	[trace]	onMoved event detected
20190506231504197	[trace]	Sync 1cogbmb (push) queued (1 waiting to sync)
20190506231504650	[trace]	Sync 1cogbmb completed (0 waiting to sync)
20190506231517765	[trace]	onCreated event detected
20190506231517806	[trace]	Sync 4vdnq5 (push) queued (1 waiting to sync)
20190506231518353	[trace]	Sync 4vdnq5 completed (0 waiting to sync)
20190506231544370	[trace]	Starting up
20190506231544418	[trace]	{"displayIntro":false,"displayPermissions":false,"lastUpdated":"2019-05-06T21:15:18.303Z","networkDisconnected":false,"syncEnabled":true,"syncId":"4717c07a4f1741889a1d32b506aba868","syncVersion":"1.5.0beta6","appVersion":"1.5.0beta6","platform":{"name":"chrome","version":"74.0.3729","mobile":false,"os":"Linux x86_64"}}
20190508172101249	[trace]	Starting up
20190508172101519	[trace]	{"displayIntro":false,"displayPermissions":false,"lastUpdated":"2019-05-06T21:15:18.303Z","networkDisconnected":false,"syncEnabled":true,"syncId":"4717c07a4f1741889a1d32b506aba868","syncVersion":"1.5.0beta6","appVersion":"1.5.0beta6","platform":{"name":"chrome","version":"74.0.3729","mobile":false,"os":"Linux x86_64"}}
20190508172104840	[trace]	Updates found, retrieving latest sync data
20190508172105293	[trace]	Sync wsxjoh (pull) queued (1 waiting to sync)
20190508172106903	[trace]	Local population completed in 1.166s
20190508172107325	[trace]	Sync wsxjoh completed (0 waiting to sync)

**Comment by ipk01** [comment]
Any news on this end? I have sync'ing turned off on all my machines, which is fine for now, but I'd love to be able to sync again and have my latest bookmarks everywhere ... ??? :-)

**Comment by nero120** [comment]
@ipk01 sorry for the delay, had no time over the past few months due to family/work commitments. Work appears to be calming down a bit this week so hoping to get some time to get these issues resolved and the beta released. Bear with me.

**Comment by nero120** [comment]
Please try again using the latest [beta](https://github.com/xbrowsersync/app/releases/tag/v1.5.0beta). I've included more logging as well as a few other fixes. If you do experience the issue again, could you please attach the xBrowserSync logs from all affected browsers?

**Comment by ipk01** [comment]
Bookmarks are garbled again. Please find the log for one of the two machines I have set-up with beta7 here:
https://www.dropbox.com/s/fv49ojwbg52dgn4/xBrowserSyncLog_20190605073926882_chrome_windwos.txt?dl=0
The other log tonight, if you then still need it.

**Comment by nero120** [comment]
Thanks for that @ipk01, I've now got a much clearer picture as to what is happening. Bear with me whilst I dig into.

**Comment by ipk01** [comment]
And just in case ... here the log from my ther machine:
https://www.dropbox.com/s/4dky28u5fjl2mk3/xBrowserSyncLog_20190605224839980_chrome_ubuntu16.txt?dl=0

**Comment by nero120** [comment]
Thanks to your logs I've identified a flaw in the sync, probably exposed by the fact that the other sync in the background was removing bookmarks at a rate that users just wouldn't do with normal behaviour. I think it's an egde case but it may explain the other sync issues that users have reported that are difficult to explain.

The fix I'm implementing will mean that syncing will be far more stable and will fail gracefully with any issues without corrupting synced data. I'm aiming to have a new beta out next week and with this fix combined with the warning to disable other syncs, hopefully sync issues will then be a thing of the past! 🤞

**Comment by ipk01** [comment]
Hi Rich,

my bookmarks garbled again ...

I have now "native" synchronization in general still on for my devices, but the booksmarks syncs disabled and just verified that. I have saved the logs from 5 environments here. The problem must have been with one of the two ubuntu browsers that produced the two .logs with the newest timestamp since on all others I fired up the browsers (at work this morning) to see my bookmarks go and come back truncated and corrupt.

https://www.dropbox.com/sh/av26gl4p3ral3a5/AAD4QhJrMh7vHHIdR_cmxe8Za?dl=0

Let me know if you can make anything out of this or if something is missing ... xbrowsersync synchronization is now disabled on all again, but I'd be happy to try once more if there is something special you would like me to watch.

Cheers

Ingo

**Comment by nero120** [comment]
@ipk01 thanks for the feedback, I believe I've cracked the sync corruption issues now once and for all. I'm currently testing a new version of syncing local bookmark changes in Chrome and have not had any corruption yet so am hopeful. I just need to implement these updates in the Firefox code and then I'll update the beta, hopefully before the end of the week.

If we can confirm then that the issues are resolved I can finally release v1.5.0! 🤞

**Comment by ipk01** [comment]
Awesome... 😁😁😁

Am 19. Juni 2019 10:52:12 MESZ schrieb Rich <notifications@github.com>:
>@ipk01 thanks for the feedback, I believe I've cracked the sync
>corruption issues now once and for all. I'm currently testing a new
>version of syncing local bookmark changes in Chrome and have not had
>any corruption yet so am hopeful. I just need to implement these
>updates in the Firefox code and then I'll update the beta, hopefully
>before the end of the week.>
>>
>If we can confirm then that the issues are resolved I can finally
>release v1.5.0! 🤞 >
>>
>-- >
>You are receiving this because you were mentioned.>
>Reply to this email directly or view it on GitHub:>
>https://github.com/xbrowsersync/app/issues/89#issuecomment-503469979

-- 
Sent from my mobile phone. Please excuse my brevity.

**Comment by nero120** [comment]
Ok here we go @ipk01: [v1.5.0beta11](https://github.com/xbrowsersync/app/releases/tag/v1.5.0beta).

I've been testing for the last couple of days syncing across multiple machines and browsers and not hit any issues yet. I'm hoping this will resolve all of your syncing issues. 🤞

Please be aware however, due to a [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1556427) in Firefox's bookmarks API, changing multiple bookmarks simultaneously will likely result in an error in xBrowserSync. The issues has been fixed in the latest [Firefox Nightly (v69.0a1)](https://www.mozilla.org/en-US/firefox/nightly/all/) so please download and test with Nightly so that you don't hit that known issue.

**Comment by ipk01** [comment]
Had this running for two days now on three machines on chrome and firefox on windows and ubuntu with no incidents so far ... Great work!

**Comment by nero120** [comment]
That's great news @ipk01, thanks so much for testing and feeding back!

In the meantime I've continued testing, found some edge case issues and fixed them. However I keep finding bugs in Firefox's bookmarks API which is driving me mad!! Mozilla also seem to take years to fix them as they deem API issues low priorty, not impressed with them at all.

I'll aim to (finally) release v1.5.0 in a week or so once I'm confident there are no other sync issues lurking. Thanks again for your help with this, you are a gentleman and a scholar!! 🙇

**Comment by ipk01** [comment]
2 weeks down the line still working great. Installed on 4 machines between various Ubuntu versions and Windows on Chrome and Firefox (where "nightly" available) and no issues at all. Syncs flawlessly now! Awesome!!!

**Comment by nero120** [comment]
Excellent @ipk01!! I've fixed a few minor issues and made some additional improvements (mostly UI) and I think I'm now ready to release v1.5.0 which should be out by the end of the week... 🤞

**Comment by nero120** [comment]
@ipk01 just to let you know I've just released [v1.5.0](https://github.com/xbrowsersync/app/releases/tag/v1.5.0) so feel to switch back from the beta.

**Comment by ipk01** [comment]
Already done that... It works really great now... The 10 pages info to click through are a lot if you have to go through them on 4 machines with 2 Browsers each... but it's OK.... Only once 😂😂😂. Love it.
Ingo

Am 1. August 2019 12:04:54 MESZ schrieb Rich <notifications@github.com>:
>@ipk01 just to let you know I've just released
>[v1.5.0](https://github.com/xbrowsersync/app/releases/tag/v1.5.0) so
>feel to switch back from the beta.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/xbrowsersync/app/issues/89#issuecomment-517218458

-- 
Sent from my mobile phone. Please excuse my brevity.

**Comment by nero120** [comment]
> The 10 pages info to click through are a lot if you have to go through them on 4 machines with 2 Browsers each

I agree absolutely, that's why I added the `skip` link on the first page! 😄 It only displays on first install so hopefully it's not too intrusive, though I get a lot of questions and issues from users who haven't read it so hopefully it will mitigate *some* of these issues.
